src/lib/listeners/ProgressEventListener.ts

```diff
@@ -0,0 +1,192 @@
+	 */
+	public constructor(display: Display, tickTime: number = 1000) {
+		super(display, tickTime);
+		this.isListening = false;
+	}
+
+	public listen(codeAnalyzer: CodeAnalyzer): void {
+		/* istanbul ignore else: should not be calling listen() twice */
+		if (!this.isListening) {
+			this.isListening = true;
+			this.engineNames = codeAnalyzer.getEngineNames();
+			this.completionPercent = 0;
+			codeAnalyzer.onEvent(EventType.RuleSelectionProgressEvent, (e: RuleSelectionProgressEvent) => this.handleEvent(e));
+		}
+	}
+
+	public stopListening(): void {
+		if (this.isListening) {
+			this.isListening = false;
+			this.stopSpinning();
+		}
+	}
+
+	private handleEvent(e: RuleSelectionProgressEvent): void {
+		if (!this.isListening) {
+			return;
+		}
+
+		if (!this.isSpinning()) {
+			this.startSpinning(getMessage(BundleName.ProgressEventListener, 'selection-spinner.action'));
+		}
+
+		this.completionPercent = e.percentComplete;
+		this.updateSpinner();
+		// Since the events this spinner listens to have one aggregated completion percentage, we can (and should) stop
+		// listening as soon as we get the 100% event, instead of waiting for the `stopListening()` method.
+		if (this.completionPercent === 100) {
+			this.stopListening();
+		}
+	}
+
+	protected createSpinnerStatus(): string {
+		return getMessage(BundleName.ProgressEventListener, 'selection-spinner.status',
+			[this.engineNames.join(', '), this.completionPercent, this.getSecondsSpentSpinning()]);
+	}
+}
+
+export class EngineRunProgressSpinner extends ProgressSpinner implements ProgressEventListener {
+	private isListening: boolean;
+	/**
+	 * Mapping from each engine's name to a number indicating its completion percentage.
+	 * @private
+	 */
+	private progressMap: Map<string, number>;
+
+	/**
+	 * @param display The display with which to show output to the user
+	 * @param tickTime A number of milliseconds to wait between automatic ticks. Defaults to 1000. Negative values disable ticking.
+	 */
+	public constructor(display: Display, tickTime: number = 1000) {
+		super(display, tickTime);
+		this.isListening = false;
+	}
+
+	public listen(codeAnalyzer: CodeAnalyzer): void {
+		/* istanbul ignore else: should not be calling listen() twice */
+		if (!this.isListening) {
+			this.isListening = true;
+			this.progressMap = new Map();
+			codeAnalyzer.onEvent(EventType.EngineRunProgressEvent, (e: EngineRunProgressEvent) => this.handleEvent(e));
+		}
+	}
+
+	public stopListening(): void {
+		if (this.isListening) {
+			this.isListening = false;
+			this.stopSpinning();
+		}
+	}
+
+	private handleEvent(e: EngineRunProgressEvent): void {
+		if (!this.isListening) {
+			return;
+		}
+
+		if (!this.isSpinning()) {
+			this.startSpinning(getMessage(BundleName.ProgressEventListener, 'execution-spinner.action'));
+		}
+
+		this.progressMap.set(e.engineName, e.percentComplete);
```

Comment:
BEHAVIORAL DIFFERENCE.
Previously, we already had a RuleSelection when we started listening, so we knew exactly which engines were going to run and could preemptively populate the map with `0`s for those entries.
Now, since we start listening much earlier, we don't actually know what engines are going to run. So we just start with an empty map and add entries for new engines as they come in. This shouldn't have an impact on the user's experience, since the engines run in parallel and Core sends a 0% completion event pretty-much immediately.
However, it means that the spinner can't auto-stop itself, because it has no way of knowing that there are no more engines incoming.